MovieTime/MovieTime/Service/MTParameterEncoding.swift

```diff
@@ -0,0 +1,375 @@
+
+    public static var prettyPrinted: ArrayEncoding { return ArrayEncoding(options: .prettyPrinted) }
+
+    public static var customQueryString: ArrayEncoding { return ArrayEncoding(customize: .reflection) }
+
+    public let options: JSONSerialization.WritingOptions
+    public let customize: Customize
+
+    public init(customize: Customize = .none, options: JSONSerialization.WritingOptions = []) {
+        self.options = options
+        self.customize = customize
+    }
+
+    public func urlEncode(_ url: String, method: MTHTTPMethod, with parameters: MTParameters?) -> URL? {
+        var url = url
+        if customize == .reflection {
+            // Encodes parameters in URL
+            if method.encodesParametersInURL() {
+                url = url.reflectionEncode(parameters: parameters)
+            }
+        }
+        guard let encodedURL = url.addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed) else { return nil }
+
+        return URL(string: encodedURL)
+    }
+
+    public func encode(_ urlRequest: URLRequest, with parameters: MTParameters?) -> URLRequest {
+        var urlRequest = urlRequest
+
+        guard let parameters = parameters, let array = parameters[MTArrayParametersKey] else {
+            return urlRequest
+        }
+        guard let method = MTHTTPMethod(rawValue: urlRequest.httpMethod ?? "GET"), !method.encodesParametersInURL() else {
+            return urlRequest
+        }
+
+        do {
+            let data = try JSONSerialization.data(withJSONObject: array, options: options)
+
+            if urlRequest.value(forHTTPHeaderField: "Content-Type") == nil {
+                urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
+            }
+
+            urlRequest.httpBody = data
+        } catch {
+            return urlRequest
+        }
+
+        return urlRequest
+    }
+}
+
+// MARK: - Custom Extensions
+
+/// Extension that allows an array be sent as a request parameters
+extension Array {
+    func asParameters() -> MTParameters {
+        return [MTArrayParametersKey: self]
+    }
+}
+
+extension NSArray {
+    func asParameters() -> MTParameters {
+        return [MTArrayParametersKey: self]
+    }
+}
+
+extension String {
+
+    func reflectionEncode(parameters: MTParameters?) -> String {
+        guard let parameters = parameters else { return self }
+
+        var strUrl = self
+
+        for (key, value) in parameters {
+
+            let strKey = String(describing: key)
+            var strValue = ""
+
+            if let array = value as? [Any] {
+
+                // For array objects
+                for (index, arrayValue) in array.enumerated() {
+
+                    let currentValue = String(describing: arrayValue)
+                    strValue = strValue.appendingFormat("%@=%@", strKey, currentValue)
+                    if index != (array.count - 1) {
+                        strValue = strValue.appending("&")
```

Comment:
is there a method in Apple's API that allows us to avoid writing our own logic for adding separators?

Reply:
Yes and I already use it in service classes:
```
appendingPathComponent(...)```
```
But in this case I'm working with the literal string and not with the URL. In addition to being a custom encode, I replace some key `{key}` in the URL:
```
let url = "http://www.google.com.br/{myKey}"

let parameters = ["myKey": "MovieTime"]

let newUrl = url.reflectionEncode(parameters: parameters) // "http://www.google.com.br/MovieTime"
```

Reply:
there's a method you've used in another part of your project: `joined(by:)`. Would it fit here as well?

Reply:
yes, I could use something like that:
```
....
let strKey = String(describing: key)
let strValue: String

if let array = value as? [Any] {
    // For array objects
    strValue = array.map {"\(strKey)=\(String(describing: $0 ))" }.joined(separator: "&")
} else {
    // For any objects
    strValue = String(describing: value)
}
....
```